src/spikeinterface/sortingcomponents/peak_detection.py

```diff
@@ -105,6 +109,220 @@ def detect_peaks(recording, method='by_channel', pipeline_nodes=None,
+        self.num_iterations = num_iterations
+
+    def get_trace_margin(self) -> int:
+        """
+        Calculate the maximum trace margin from the internal pipeline.
+        Using the strategy as use by the Node pipeline
+
+
+        Returns
+        -------
+        int
+            The maximum trace margin.
+        """
+        internal_pipeline = (self.peak_detector_node, self.waveform_extraction_node, self.waveform_denoising_node)
+        pipeline_margin = (node.get_trace_margin() for node in internal_pipeline if hasattr(node, "get_trace_margin"))
+        return max(pipeline_margin)
+
+    def compute(self, traces_chunk, start_frame, end_frame, segment_index, max_margin) -> Tuple[np.ndarray]:
+        """
+        Perform the iterative peak detection, waveform extraction, and denoising.
+
+        Parameters
+        ----------
+        traces_chunk : array-like
+            The chunk of traces to process.
+        start_frame : int
+            The starting frame for the chunk.
+        end_frame : int
+            The ending frame for the chunk.
+        segment_index : int
+            The segment index.
+        max_margin : int
+            The maximum margin for the traces.
+
+        Returns
+        -------
+        tuple of ndarray
+            A tuple containing a single ndarray with the detected peaks.
+        """
+
+        traces_chunk = traces_chunk.astype("float32")
+        local_peaks_list = []
+        all_waveforms = []
+
+        for iteration in range(self.num_iterations):
+            (local_peaks,) = self.peak_detector_node.compute(
+                traces=traces_chunk, 
+                start_frame=start_frame, 
+                end_frame=end_frame,
+                segment_index=segment_index,
+                max_margin=max_margin,
+            )
+
+            local_peaks = self.add_iteration_to_peaks_dtype(local_peaks=local_peaks, iteration=iteration)
+            local_peaks_list.append(local_peaks)
+
+            # End algorith if no peak is found
+            if local_peaks.size == 0:
+                break
+
+            waveforms = self.waveform_extraction_node.compute(traces=traces_chunk, peaks=local_peaks)
+            denoised_waveforms = self.waveform_denoising_node.compute(
+                traces=traces_chunk, 
+                peaks=local_peaks, 
+                waveforms=waveforms
+            )
+            traces_chunk_minus_waveforms, _ = self.substract_waveforms_from_traces(
+                traces_chunk, local_peaks, 
+                denoised_waveforms
+            )
+
+            # update traces_chunk to run the algorithm again and store local_peaks
+            traces_chunk = traces_chunk_minus_waveforms
+
+            all_waveforms.append(waveforms)
+
+        all_local_peaks = np.concatenate(local_peaks_list, axis=0)
+        all_waveforms = np.concatenate(all_waveforms, axis=0) if len(all_waveforms) != 0 else np.empty((0, 0, 0))
+
```

Comment:
Maybe we should order peak by there sample_index and apply this order also to waveforms.

Reply:
This is going to have the same order as the base method (i.e. `self.peak_detector_node`). Why we would order this if the base method is not ordered? Should all the methods be ordered? How are you thinking about?

Reply:
in this case because this is iterative you can find peak on the second that are earlier that the last one.

Reply:
Good point.

Reply:
Done in the last commit. I added a test specifically for this and then changed the code for passing it.